README.md
### Maker collaterals monitoring bot

The bot is a Prometheus exporter which parses loan data from Maker protocol and calculates risk distribution based on
collateral to loan ratio, see [zones definition](#zones-definition) below.

#### Run

`docker compose`

- Create a `.env` file from `.env.example` and fill in the required variables.
- Execute command:

```bash
docker compose up -d bot
```

`docker`

- Create a `.env` file from `.env.example` and fill in the required variables.
- Build image:

```bash
docker build -t maker-risks-bot .
```

- Run container:

```bash
docker run -d -P --env-file ./.env maker-risks-bot
```

`dev`

- Expose environment variables presented at `.env.example`.
- Install dependencies via poetry:

```bash
poetry install
```

- Execute command:

```bash
python src/main.py
```

#### Zones definition

Risk zones are defined as ranges of collateral to loan ration

| Zone        | >    | <    |
| ----------- | ---- | ---- |
| A           | 2.50 | ∞    |
| B+          | 1.75 | 2.50 |
| B           | 1.50 | 1.75 |
| B-          | 1.25 | 1.50 |
| C           | 1.10 | 1.25 |
| D           | 1.00 | 1.10 |
| Liquidation | 0    | 1.00 |

#### The most essential exposed metrics

- `{}_collateral_percentage{ilk=<ilk>, zone=<zone>}` is computed percent of collaterals in the given zone
- `{}_collateral_value{ilk=<ilk>, zone=<zone>}` is computed value of collaterals in the given zone
- `{}_processing_finished_seconds{ilk=<ilk>` is timestamp of the last metrics update
- `{}_bot_last_block_num` is the last block parser results was updated against
- `{}_eth_latest_block_num` is the latest block available for the parser

#### Visualization

Pre-built Grafana dashboards are available in the `./grafana` directory. The dashboards use Prometheus as a source of
data. To run via docker compose use: `docker compose up -d bot prometheus grafana`

#### Alerting

The exporter was built primarily for alerting purposes. Therefore, in the `./prometheus` directory, predefined rules and
the related tests are located. `status.rule` is created for monitoring the status of the exporter. `zones_changes.rule`
defines rules for monitoring collaterals.

docker-compose configuration allows running the setup for alerting via Alertmanager by sending alerts to Discord.
Configure webhook parameters in `./alertmanager-discord/alertmanager-discord.yml` and start the full docker-compose
stack via command `docker compose up -d`.

## Release flow

To create a new release:

1. Merge all changes to the `master` branch.
1. After the merge, the `Prepare release draft` action will run automatically. When the action is complete, a release
   draft is created.
1. When you need to release, go to Repo → Releases.
1. Publish the desired release draft manually by clicking the edit button - this release is now the `Latest Published`.
1. After publication, the action to create a release bump will be triggered automatically.


CHANGELOG.md
# This file is no longer used by semantic-release
For the latest release notes, please see the [GitHub releases
page](https://github.com/lidofinance/maker-risks-bot/releases).



## [1.10.1](https://github.com/lidofinance/maker-risks-bot/compare/1.10.0...1.10.1) (2023-04-11)


### Bug Fixes

* allow retry on null transaction receipt ([47aac4f](https://github.com/lidofinance/maker-risks-bot/commit/47aac4fe89ec2e51d6f434ac0463473fc8c33041))
* use getLogs instead of createFilter ([fdf8258](https://github.com/lidofinance/maker-risks-bot/commit/fdf825839208a9ae3675a1b22ffb41e8ee419f20))



# [1.10.0](https://github.com/lidofinance/maker-risks-bot/compare/1.9.0...1.10.0) (2023-01-30)


### Features

* expose provider to RPC metrics ([3d918e8](https://github.com/lidofinance/maker-risks-bot/commit/3d918e8c408cad8e809b3bc7fb8b4b0e93102121))
* expose provider to RPC metrics ([2004764](https://github.com/lidofinance/maker-risks-bot/commit/200476444104ecaabf3e8f5f57dbb96d9a122cd4))



# [1.9.0](https://github.com/lidofinance/maker-risks-bot/compare/1.8.0...1.9.0) (2023-01-12)


### Bug Fixes

* add fallback for eth provider ([c5cad3d](https://github.com/lidofinance/maker-risks-bot/commit/c5cad3d9fc5fab7d93a35a7b66fba8eb832b0f6b))
* division by zero ([0bc90b4](https://github.com/lidofinance/maker-risks-bot/commit/0bc90b4662a8b8a317aa6c047d007166fae8bf49))
* fix main cycle's error handle ([9245d7b](https://github.com/lidofinance/maker-risks-bot/commit/9245d7bc6d90bc756fc8893ef46585cf7ad05041))
* remove unused computation ([0d9bac6](https://github.com/lidofinance/maker-risks-bot/commit/0d9bac6b1eef5fed09ce5b6e65f8f3623924fdf4))
* upd rules ([3f7c60d](https://github.com/lidofinance/maker-risks-bot/commit/3f7c60dc2f5af7cf656fffb762c9e4fce9b41b6e))
* use loan value round ([bb24b88](https://github.com/lidofinance/maker-risks-bot/commit/bb24b88dfdfce666cfd2286e46dcb442179735bd))


### Features

* add retryable middleware ([a90c8f3](https://github.com/lidofinance/maker-risks-bot/commit/a90c8f352a76512cbe0a6c0c0af385907daa1608))
* expose loans value to alerts ([56d7fff](https://github.com/lidofinance/maker-risks-bot/commit/56d7fff2e399e1d1defd3c5b91c7dd9c0d038de0))
* rethink main loop ([b5cc629](https://github.com/lidofinance/maker-risks-bot/commit/b5cc6295b89da4aa93ee60dbb6d557337bb7e422))
* update middleware ([5778d40](https://github.com/lidofinance/maker-risks-bot/commit/5778d40380efb254fe7d46cfd99eb067dc9a6bc3))



# [1.8.0](https://github.com/lidofinance/maker-risks-bot/compare/1.7.0...1.8.0) (2022-09-29)


### Bug Fixes

* upd rules ([e06aee9](https://github.com/lidofinance/maker-risks-bot/commit/e06aee9969f1b2aa5344e62df98c6b01d5762441))


### Features

* expose loans value to alerts ([6521035](https://github.com/lidofinance/maker-risks-bot/commit/6521035e56c3ac8079bf2a4f62354893d01e09e7))



# [1.7.0](https://github.com/lidofinance/maker-risks-bot/compare/1.6.0...1.7.0) (2022-09-15)


### Bug Fixes

* **grafana:** fix timestamps metrics panels ([0ed067b](https://github.com/lidofinance/maker-risks-bot/commit/0ed067be39650d252eb89c55733d4975684f2ec3))
* passthrough log level ([ebcdd1f](https://github.com/lidofinance/maker-risks-bot/commit/ebcdd1fb19b21b502150e86c81c813fc669044b7))
* **prometheus:** update rules thresholds ([515e444](https://github.com/lidofinance/maker-risks-bot/commit/515e44490aa025313617de1e6aa750d8a010b38f))
* update curl version ([0feef9f](https://github.com/lidofinance/maker-risks-bot/commit/0feef9f67251fde6cd6d6bbd06755f69bb998093))
* update Dockerfile for the latest Debian ([3773c63](https://github.com/lidofinance/maker-risks-bot/commit/3773c631b1220563a5f9edcfdcdaefc315514004))


### Features

* add build-info parsing ([c3ed854](https://github.com/lidofinance/maker-risks-bot/commit/c3ed854a6f06ee53044931f6b15065fff8503e39))
* **grafana:** udpated blocks related panels ([a5bce64](https://github.com/lidofinance/maker-risks-bot/commit/a5bce64102194f53bfd56aa56721056cd05c0a68))
* **metrics:** improve status metrics and update prometheus rules ([e873605](https://github.com/lidofinance/maker-risks-bot/commit/e8736058561fec59a1661d3f035179ea637b868d))
* parse data onchain ([#37](https://github.com/lidofinance/maker-risks-bot/issues/37)) ([91b83f2](https://github.com/lidofinance/maker-risks-bot/commit/91b83f23a8e8d6ef0a9aa4a7af0374bb3be3fa13))
* remove Maker DataAPI parser ([aba077c](https://github.com/lidofinance/maker-risks-bot/commit/aba077c84be4501d08571c1a98c67c7df121e5a7))



# [1.6.0](https://github.com/lidofinance/maker-risks-bot/compare/1.5.0...1.6.0) (2022-08-11)


### Bug Fixes

* **grafana:** fix timestamps metrics panels ([ea1ab4f](https://github.com/lidofinance/maker-risks-bot/commit/ea1ab4fcce0ae9999267ce30437d64973f64a558))
* update curl version ([0006ebb](https://github.com/lidofinance/maker-risks-bot/commit/0006ebbffded4161f95ac0ac36196c3eab9e6989))


### Features

* **metrics:** improve status metrics and update prometheus rules ([e9374ec](https://github.com/lidofinance/maker-risks-bot/commit/e9374ecd4d854e6f2b8d9259f85a6d3d7bb16fa3))



# [1.5.0](https://github.com/lidofinance/maker-risks-bot/compare/1.4.0...1.5.0) (2022-07-29)


### Bug Fixes

* CI workflows ([1175abb](https://github.com/lidofinance/maker-risks-bot/commit/1175abb27a89f17b4dc7791c82350dfff0617ff0))
* **grafana:** one more datasource fix ([#8](https://github.com/lidofinance/maker-risks-bot/issues/8)) ([414aecf](https://github.com/lidofinance/maker-risks-bot/commit/414aecf9c0846cbffe2c7a56a96ae1534a0b16a9))
* retrieve token on demand only ([1cef262](https://github.com/lidofinance/maker-risks-bot/commit/1cef262e96fb2c17f64d40c1cd65e2ccde6b48e0))
* testnet deploy ([dd496ca](https://github.com/lidofinance/maker-risks-bot/commit/dd496ca27ed5bc6c9e72094dfb8526ce1ef86371))


### Features

* **grafana:** add status dashboard ([afdfa3f](https://github.com/lidofinance/maker-risks-bot/commit/afdfa3f2ca48bfda1fbb480712ac998e0b4f1f55))

src/bot/analytics.py
"""Analytics methods"""

from contextlib import suppress
from typing import List

import pandas as pd

from .consts import RAY_DECIMALS
from .ilks import MakerIlk
from .parsers import BaseParser

RISK_LABELS = ["A", "B+", "B", "B-", "C", "D", "liquidation"]
RISK_VALUES = [2.50, 1.75, 1.50, 1.25, 1.10, 1.00]


def prepare_data(df: pd.DataFrame, asset: MakerIlk, parser: BaseParser) -> pd.DataFrame:
    """Transform raw data received"""

    df = df.copy()

    df["ink"] = df["ink"] / pow(10, asset.decimals)
    df["art"] = df["art"] / pow(10, asset.decimals)
    df.rename(columns={"art": "debt", "ink": "collateral"}, inplace=True)

    df.fillna(0, inplace=True)
    df = pd.DataFrame(data=df.query("collateral > 0 and debt > 0"))

    # rate => stablecoin debt multiplier (e.g. 1.015)
    # spot => maximum stablecoin allowed per unit of collateral (e.g. 1889.2)
    (_, rate, spot, _, _) = parser.get_vat_stats(asset)
    rate = rate / pow(10, RAY_DECIMALS)
    spot = spot / pow(10, RAY_DECIMALS)

    df["healthf"] = df["collateral"] * spot / (df["debt"] * rate)

    return df


def get_risks(df: pd.DataFrame, ratio_list: List[float]) -> pd.DataFrame:
    """
    This function calculates the risk level for each position
    and returns the positions sorted by risk
    """

    df = df.copy()

    df["risk_rating"] = [
        (x > ratio_list[0] and "A")
        or (ratio_list[1] < x <= ratio_list[0] and "B+")
        or (ratio_list[2] < x <= ratio_list[1] and "B")
        or (ratio_list[3] < x <= ratio_list[2] and "B-")
        or (ratio_list[4] < x <= ratio_list[3] and "C")
        or (ratio_list[5] < x <= ratio_list[4] and "D")
        or (ratio_list[5] <= x and "liquidation")
        for x in df["healthf"]
    ]

    df.sort_values(by="healthf", ascending=False, inplace=True)

    return df


def get_distr(data) -> pd.DataFrame:
    """This function calculates and returns a pivot table by risk levels"""

    risk_distr = data.pivot_table(index="risk_rating", values=["collateral"], aggfunc=["sum"])
    risk_distr.columns = ["ilk"]
    risk_distr["percent"] = (risk_distr["ilk"] / risk_distr["ilk"].sum()) * 100

    return risk_distr


def calculate_values(
    data: pd.DataFrame,
    asset: MakerIlk,
    parser: BaseParser,
) -> dict[str, dict]:
    """Calculate risk distribution.
    Almost as is from related jupyter notebook."""

    df = prepare_data(data, asset, parser)

    df = get_risks(df, RISK_VALUES)
    risk_distr = get_distr(df)

    values = {}
    for label in RISK_LABELS:
        value = {
            "percent": 0,
            "ilk": 0,
        }

        with suppress(KeyError):
            value = risk_distr.loc[label].to_dict()

        values[label] = value

    return values
